RealmSwift/SwiftUI.swift

```diff
@@ -643,6 +647,368 @@ extension EnvironmentValues {
             self[RealmEnvironmentKey.self] = newValue.configuration
         }
     }
+    /// The current `PartitionValue` that the view should use.
+    public var partitionValue: PartitionValue? {
+        get {
+            return self[PartitionValueEnvironmentKey.self]
+        }
+        set {
+            self[PartitionValueEnvironmentKey.self] = newValue
+        }
+    }
+}
+
+@available(iOS 14.0, macOS 11.0, tvOS 14.0, watchOS 7.0, *)
+private class ObservableAsyncOpenStorage: ObservableObject {
+    private var app: App
+    var configuration: Realm.Configuration
+    var partitionValue: AnyBSON
+
+    var cancellables = [AnyCancellable]()
+
+    @Published var asyncOpenState: AsyncOpenState = .connecting {
+        willSet {
+            objectWillChange.send()
+        }
+    }
+
+    func asyncOpen() -> AnyPublisher<RealmPublishers.AsyncOpenPublisher.Output, RealmPublishers.AsyncOpenPublisher.Failure> {
+        if let currentUser = app.currentUser,
+           currentUser.isLoggedIn {
+            return asyncOpenForUser(app.currentUser!, partitionValue: partitionValue, configuration: configuration)
+        } else {
+            asyncOpenState = .waitingForUser
+            return app.objectWillChange
+                .filter { $0.currentUser != nil }
```

Comment:
Can probably use `compactMap` here:
`.compactMap { $0.currentUser }`
to make the code simpler/more legible.

Reply:
Or even better:
`.compactMap(\.currentUser)`

Reply:
Pretty nice, thanks